### PR TITLE
User/anramanath/process data element array

### LINKED ISSUE #110: Optimize DataCooker processing path
Update BaseSourceProcessingSession<T, TContext, TKey>.ProcessDataElement to use a for loop rather than a foreach loop, and to use an array rather than an IEnumerable.

### PR DESCRIPTION
**Pull request by anramanath**
Making ProcessDataElement more efficient. fixes #110.

### CODE REVIEW

File: src/Microsoft.Performance.SDK.Runtime/Extensibility/BaseSourceProcessingSession.cs [hunk 1]
```diff
@@ -1,16 +1,16 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
-using System;
```
Comment (mslukebo) [inline src/Microsoft.Performance.SDK.Runtime/Extensibility/BaseSourceProcessingSession.cs:4]: I think we're deciding to keep System imports first. You can change this behavior in VS Options under Text Editor -> C# -> Advanced

File: src/Microsoft.Performance.SDK.Runtime/Extensibility/BaseSourceProcessingSession.cs [hunk 2]
```diff
@@ -116,17 +116,18 @@ public DataProcessingResult ProcessDataElement(T data, TContext context, Cancell
 
             var key = data.GetKey();
 
-            if (!this.activeCookerRegistry.ContainsKey(key) && !this.receiveAllDataCookers.Any())
+            if (!this.activeCookerRegistry.ContainsKey(key) && this.receiveAllDataCookers.Count == 0)
```
Comment (mslukebo) [inline src/Microsoft.Performance.SDK.Runtime/Extensibility/BaseSourceProcessingSession.cs:119]: why this change?

File: src/Microsoft.Performance.SDK.Runtime/Extensibility/BaseSourceProcessingSession.cs [hunk 3]
```diff
@@ -143,17 +144,18 @@ DataProcessingResult ProcessDataCookers(IEnumerable<ISourceDataCooker<T, TContex
...
-                var sourceDataCookers = this.activeCookerRegistry[key];
                 keyedResult = ProcessDataCookers(sourceDataCookers);
                 if (keyedResult == DataProcessingResult.CorruptData)
                 {
                     return keyedResult;
                 }
             }
 
-            var allDataResult = ProcessDataCookers(this.activeReceiveAllDataCookers);
+            var allDataCookers = this.activeReceiveAllDataCookers;
```
Comment (mslukebo) [inline src/Microsoft.Performance.SDK.Runtime/Extensibility/BaseSourceProcessingSession.cs:157]: un-needed variable

File: src/Microsoft.Performance.SDK.Runtime/Extensibility/BaseSourceProcessingSession.cs [hunk 4]
```diff
@@ -190,15 +192,16 @@ protected void InitializeForSourceParsing()
...
         }
 
         protected void ActivateCooker(ISourceDataCooker<T, TContext, TKey> sourceDataCooker)
         {
             Guard.NotNull(sourceDataCooker, nameof(sourceDataCooker));
 
             this.activeCookers.Add(sourceDataCooker);
-            if (sourceDataCooker.Options.HasFlag(SourceDataCookerOptions.ReceiveAllDataElements))
+            if (sourceDataCooker.Options.HasFlag(SourceDataCookerOptions.ReceiveAllDataElements) &&
+                !this.activeReceiveAllDataCookers.Contains(sourceDataCooker))
```
Comment (mslukebo) [inline src/Microsoft.Performance.SDK.Runtime/Extensibility/BaseSourceProcessingSession.cs:204]: this might be expensive. is this a hot path?

**Review by mslukebo — CHANGES_REQUESTED** [review]

File: src/Microsoft.Performance.SDK.Runtime/Extensibility/BaseSourceProcessingSession.cs (hunk 2 above)
Comment (anramanath) [reply src/Microsoft.Performance.SDK.Runtime/Extensibility/BaseSourceProcessingSession.cs:119]: Using .Any() requires an enumerator. Since Count is cached, we can access that faster.

File: src/Microsoft.Performance.SDK.Runtime/Extensibility/BaseSourceProcessingSession.cs (hunk 4 above)
Comment (anramanath) [reply src/Microsoft.Performance.SDK.Runtime/Extensibility/BaseSourceProcessingSession.cs:204]: It could be expensive, but should not be a problem as it isn't a hot path.

**Review by mslukebo — APPROVED** [review]

**Review by zachnew-msft — APPROVED** [review]

**Review by mslukebo — APPROVED** [review]

**Review by zachnew-msft — APPROVED** [review]
